### PR TITLE
feat: impl MetaStorage interface

### LINKED ISSUE #15: feat: impl MetaStorage with etcd kv
**Description**

impl MetaStorage with etcd kv

**Proposal**
refer to https://github.com/CeresDB/ceresmeta/blob/615113cf1bb0ee1f002dd0448194f4aed7cb3e99/server/storage/storage_impl.go#L26

**Additional context**

### PR DESCRIPTION
**Pull request by baojinri**
# Which issue does this PR close?

Closes #15 

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

# How does this change test

### CODE REVIEW

File: server/storage/key_path.go [hunk 1]
```diff
@@ -10,6 +10,11 @@ import (
 const (
 	cluster = "v1/cluster"
 	schema  = "schema"
+	node    = "node"
+	table   = "table"
+	shard   = "shard"
+	topo    = "topo"
+	lv      = "latest_version"
```
Comment (chunshao90) [inline server/storage/key_path.go:17]: latest_version

File: server/storage/key_path.go [hunk 2]
```diff
@@ -10,6 +10,11 @@ import (
 const (
 	cluster = "v1/cluster"
 	schema  = "schema"
+	node    = "node"
+	table   = "table"
+	shard   = "shard"
+	topo    = "topo"
```
Comment (chunshao90) [inline server/storage/key_path.go:16]: topology

File: server/storage/key_path.go [hunk 3]
```diff
@@ -20,3 +25,31 @@ const (
...
+
+func makeShardKey(clusterID uint32, shardID uint32, latestVersion string) string {
+	return path.Join(cluster, fmt.Sprintf("%020d", clusterID), shard, fmt.Sprintf("%020d", shardID), latestVersion)
+}
+
+func makeClusterTopologyKey(clusterID uint32, latestVersion string) string {
+	return path.Join(cluster, fmt.Sprintf("%020d", clusterID), topo, latestVersion)
+}
+
+func makeLatestVersion(clusterID uint32) string {
```
Comment (chunshao90) [inline server/storage/key_path.go:49]: makeClusterTopologyLatestVersion

File: server/storage/storage_test.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/server/v3/embed"
+)
+
+func TestCluster(t *testing.T) {
+	re := require.New(t)
+	s := NewStorage(t)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultRequestTimeout)
+	defer cancel()
+	meta := &metapb.Cluster{Id: 0, Name: "name", MinNodeCount: 0, ReplicationFactor: 0, ShardTotal: 0}
```
Comment (chunshao90) [inline server/storage/storage_test.go:24]: const clusterID: 100
const clusterName: "name"
...

File: server/storage/storage_impl.go [hunk 5]
```diff
@@ -45,18 +46,64 @@ func newEtcdStorage(client *clientv3.Client, rootPath string, opts Options) *Met
...
 func (s *MetaStorageImpl) GetCluster(ctx context.Context, clusterID uint32) (*metapb.Cluster, error) {
-	return nil, nil
+	key := makeClusterKey(clusterID)
+	value, err := s.Get(ctx, key)
+	if err != nil {
+		return nil, errors.Wrapf(err, "get cluster err, key: %v", key)
+	}
+	meta := &metapb.Cluster{}
+	if err = proto.Unmarshal([]byte(value), meta); err != nil {
+		return nil, ErrMetaGetCluster.WithCausef("proto parse err: %v", err)
```
Comment (ShiKaiWi) [inline server/storage/storage_impl.go:56]: - A more accurate error is better, e.g. ErrParseClusterPb.
- No space between `:` and `%v`.
- The message may be better if it reads: `proto parse failed, err:%v`.

Besides this, some other errors in this PR should be refactored too.

File: server/storage/storage_impl.go [hunk 6]
```diff
@@ -45,18 +46,64 @@ func newEtcdStorage(client *clientv3.Client, rootPath string, opts Options) *Met
...
+	}
+	clusterMetaData := &metapb.ClusterTopology{}
+	if err = proto.Unmarshal([]byte(value), clusterMetaData); err != nil {
+		return nil, ErrMetaGetClusterTopology.WithCausef("proto parse err: %v", err)
+	}
+	return clusterMetaData, nil
 }
 
 func (s *MetaStorageImpl) PutClusterTopology(ctx context.Context, clusterID uint32, clusterMetaData *metapb.ClusterTopology) error {
+	key := makeLatestVersion(clusterID)
```
Comment (ShiKaiWi) [inline server/storage/storage_impl.go:93]: Should a new `Version` be selected to do this topology update?

File: server/storage/storage_impl.go [hunk 7]
```diff
@@ -101,33 +148,151 @@ func (s *MetaStorageImpl) ListSchemas(ctx context.Context, clusterID uint32) ([]
...
+		if err = proto.Unmarshal([]byte(value), tableData); err != nil {
+			return nil, ErrMetaGetTables.WithCausef("proto parse err: %v", err)
+		}
+		tables = append(tables, tableData)
+	}
+	return tables, nil
 }
 
 func (s *MetaStorageImpl) PutTables(ctx context.Context, clusterID uint32, schemaID uint32, tables []*metapb.Table) error {
+	for _, item := range tables {
```
Comment (ShiKaiWi) [inline server/storage/storage_impl.go:182]: Putting all the tables in a batch may be more elegant and effective.

File: server/storage/storage_impl.go [hunk 8]
```diff
@@ -101,33 +148,151 @@ func (s *MetaStorageImpl) ListSchemas(ctx context.Context, clusterID uint32) ([]
...
+		}
+		if err = s.Put(ctx, key, string(value)); err != nil {
+			return errors.Wrapf(err, "put tables err, key: %v", key)
+		}
+	}
 	return nil
 }
 
 func (s *MetaStorageImpl) DeleteTables(ctx context.Context, clusterID uint32, schemaID uint32, tableIDs []uint64) error {
+	for _, item := range tableIDs {
```
Comment (ShiKaiWi) [inline server/storage/storage_impl.go:196]: Delete all the tables in a batch may be more elegant and effective.

File: server/storage/storage_impl.go (hunk 7 above)
Comment (baojinri) [reply server/storage/storage_impl.go:193]: Todo, relating to #24

File: server/storage/storage_impl.go (hunk 8 above)
Comment (baojinri) [reply server/storage/storage_impl.go:208]: Todo, relating to https://github.com/CeresDB/ceresmeta/issues/24

File: server/storage/storage_impl.go (hunk 5 above)
Comment (ShiKaiWi) [inline server/storage/storage_impl.go]: Remove the space between `err:` and `%v`.

File: server/storage/storage_impl.go (hunk 7 above)
Comment (ShiKaiWi) [reply server/storage/storage_impl.go:193]: If the batch put is not planned to be done in this PR, maybe we should add a TODO for that.

File: server/storage/storage_impl.go [hunk 9]
```diff
@@ -27,36 +29,94 @@ type MetaStorageImpl struct {
...
+
+	rootPath string
 }
 
 // NewMetaStorageImpl creates a new base storage endpoint with the given KV and encryption key manager.
 // It should be embedded insIDe a storage backend.
 func NewMetaStorageImpl(
 	kv KV,
-	opts Options,
+	opts Options, rootPath string,
```
Comment (chunshao90) [inline server/storage/storage_impl.go:40]: one line？

File: server/storage/storage_impl.go [hunk 10]
```diff
@@ -27,36 +29,94 @@ type MetaStorageImpl struct {
 	KV
 
 	opts Options
+
+	rootPath string
```
Comment (chunshao90) [inline server/storage/storage_impl.go:33]: add rootPath field in Options?

File: server/storage/storage_impl.go [hunk 11]
```diff
@@ -27,36 +29,94 @@ type MetaStorageImpl struct {
...
-		NewEtcdKV(client, rootPath), opts)
+		NewEtcdKV(client, rootPath), opts, rootPath)
 }
 
 func (s *MetaStorageImpl) GetCluster(ctx context.Context, clusterID uint32) (*metapb.Cluster, error) {
-	return nil, nil
+	key := makeClusterKey(clusterID)
+	value, err := s.Get(ctx, key)
+	if err != nil {
+		return nil, errors.Wrapf(err, "get cluster failed, key:%v", key)
```
Comment (chunshao90) [inline server/storage/storage_impl.go:55]: meta storage get cluster failed, key:%s, clusterID:%d

File: server/storage/storage_impl.go [hunk 12]
```diff
@@ -27,36 +29,94 @@ type MetaStorageImpl struct {
...
+	}
+	meta := &metapb.Cluster{}
+	if err = proto.Unmarshal([]byte(value), meta); err != nil {
+		return nil, ErrParseGetCluster.WithCausef("proto parse failed, err:%v", err)
```
Comment (chunshao90) [inline server/storage/storage_impl.go:59]: add clusterID

File: server/storage/storage_impl.go [hunk 13]
```diff
@@ -73,7 +133,7 @@ func (s *MetaStorageImpl) ListSchemas(ctx context.Context, clusterID uint32) ([]
 			if rangeLimit /= 2; rangeLimit >= s.opts.MinScanLimit {
 				continue
 			}
-			return nil, err
+			return nil, errors.Wrapf(err, "get schemas failed, start key:%v, end key:%v, range limit:%v", startKey, endKey, rangeLimit)
```
Comment (chunshao90) [inline server/storage/storage_impl.go:136]: list schemas

File: server/storage/storage_impl.go [hunk 14]
```diff
@@ -100,34 +160,170 @@ func (s *MetaStorageImpl) ListSchemas(ctx context.Context, clusterID uint32) ([]
...
 func (s *MetaStorageImpl) ListTables(ctx context.Context, clusterID uint32, schemaID uint32, tableID []uint64) ([]*metapb.Table, error) {
-	return nil, nil
+	tables := make([]*metapb.Table, 0)
+	for _, item := range tableID {
+		key := makeTableKey(clusterID, schemaID, item)
+		value, err := s.Get(ctx, key)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get tables failed, key:%v", key)
+		}
+		tableData := &metapb.Table{}
```
Comment (chunshao90) [inline server/storage/storage_impl.go:187]: add newline above

File: server/storage/storage_impl.go [hunk 15]
```diff
@@ -20,47 +22,109 @@ type Options struct {
...
+		return nil, ErrParseGetClusterTopology.WithCausef("proto parse failed, err:%s, clusterID:%d", err, clusterID)
+	}
+	return clusterMetaData, nil
 }
 
-func (s *MetaStorageImpl) PutClusterTopology(ctx context.Context, clusterID uint32, clusterMetaData *metapb.ClusterTopology) error {
+func (s *metaStorageImpl) PutClusterTopology(ctx context.Context, clusterID uint32, latestVersion uint32, clusterMetaData *metapb.ClusterTopology) error {
+	value, err := proto.Marshal(clusterMetaData)
+	if err != nil {
+		return ErrParsePutClusterTopology.WithCausef("proto parse failed, err:%s, clusterID:%d", err, clusterID)
```
Comment (chunshao90) [inline server/storage/storage_impl.go:103]: clusterID:%d, err:%v

File: server/storage/storage_impl.go [hunk 16]
```diff
@@ -20,47 +22,109 @@ type Options struct {
...
+	cmp := clientv3.Compare(clientv3.Value(latestVersionKey), "=", fmtID(uint64(latestVersion)))
+	opPutClusterTopology := clientv3.OpPut(key, string(value))
+	opPutLatestVersion := clientv3.OpPut(latestVersionKey, fmtID(clusterMetaData.DataVersion))
+
+	resp, err := s.Txn(ctx).
+		If(cmp).
+		Then(opPutClusterTopology, opPutLatestVersion).
+		Commit()
+	if err != nil {
+		return errors.Wrapf(err, "meta storage put cluster topology failed, key:%s, clusterID:%d", key, clusterID)
```
Comment (chunshao90) [inline server/storage/storage_impl.go:118]: clusterID:%d, key:%s

File: server/storage/key_path.go [hunk 17]
```diff
@@ -18,5 +23,70 @@ const (
...
+// makeClusterTopologyKey returns the cluster topology meta info key path with the given region ID.
+// example:
+// cluster 1: v1/cluster/1/topo/1 -> ceresmeta.ClusterTopology
+//            v1/cluster/1/topo/2 -> ceresmeta.ClusterTopology
+//            v1/cluster/1/topo/3 -> ceresmeta.ClusterTopology
+func makeClusterTopologyKey(clusterID uint32, latestVersion string) string {
+	return path.Join(cluster, fmtID(uint64(clusterID)), topology, latestVersion)
+}
+
+// makeClusterTopologyLatestVersionKey returns the cluster topology latest version info key path with the given region ID.
```
Comment (chunshao90) [inline server/storage/key_path.go:73]: comment is not correct

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Review by chunshao90 — APPROVED** [review]
LGTM

**Pull request merged**